### PR TITLE
[main][feat] Revenue Treasury enhancement

### PR DESCRIPTION
**Pull request by alpanaca**
## Description
Add more flexibility and reusability for Revenue Treasury

## Why?
Previous version of the contract would need to be redeployed as
- There is not swap to transfer to vault
- There is no setter to update token, vault, remaining

## ChangeLogs:
- add setters
- add swap functionality before transfer to vault

### CODE REVIEW

File: contracts/8.10/protocol/RevenueTreasury.sol [hunk 1]
```diff
@@ -104,28 +107,73 @@ contract RevenueTreasury is Initializable, OwnableUpgradeable {
 
   /// @notice Split fund and distribute
   function feedGrassHouse() external {
```
Comment (spicysquid168) [inline contracts/8.10/protocol/RevenueTreasury.sol:109]: nonReentrant

File: contracts/8.10/protocol/RevenueTreasury.sol [hunk 2]
```diff
@@ -104,28 +107,73 @@ contract RevenueTreasury is Initializable, OwnableUpgradeable {
...
-      uint256 split = (token.myBalance() * splitBps) / 10000;
-      // The amount to transfer to vault shoule be equal to min(split , remaining)
-      _transferAmount = split < remaining ? split : remaining;
+      uint256 _split = (token.myBalance() * splitBps) / 10000;
+      // The amount to transfer to vault should be equal to min(split , remaining)
+
+      if (vaultSwapPath.length != 0) {
+        // find the amount in if we're going to cover all remaining
+        uint256[] memory expectedAmountsIn = router.getAmountsIn(remaining, vaultSwapPath);
+        // if the exepected amount in < _split, then swap with _split amount
```
Comment (spicysquid168) [inline contracts/8.10/protocol/RevenueTreasury.sol:123]: Wrong comment

Comment (spicysquid168) [reply contracts/8.10/protocol/RevenueTreasury.sol:123]: Should be
```
// if the expected amount in < _split, then swap with expected amount in
```

File: contracts/8.10/protocol/RevenueTreasury.sol [hunk 3]
```diff
@@ -104,28 +107,73 @@ contract RevenueTreasury is Initializable, OwnableUpgradeable {
...
+      token.safeApprove(address(router), _swapAmount);
+      router.swapExactTokensForTokens(_swapAmount, 0, rewardPath, address(this), block.timestamp);
+    }
 
     // Feed all reward token to grasshouse
     uint256 _feedAmount = grasshouseToken.myBalance();
     grasshouseToken.safeApprove(address(grassHouse), _feedAmount);
     grassHouse.feed(_feedAmount);
-    emit LogFeedGrassHouse(msg.sender, _transferAmount, _swapAmount, _feedAmount);
+    emit LogFeedGrassHouse(msg.sender, _transferAmount, _feedAmount);
```
Comment (spicysquid168) [inline contracts/8.10/protocol/RevenueTreasury.sol:157]: LogSettleBadDebt as well

Comment (alpanaca) [reply contracts/8.10/protocol/RevenueTreasury.sol:157]: LogSettleBadDebt(_trasnferAmount)  as a separate log?

**Review by jr-alpaca — APPROVED** [review]

**Review by abc-alpaca — APPROVED** [review]

**Pull request merged**